doc/proposals/helm-operator.md

```diff
@@ -0,0 +1,127 @@
+
+### New Operator Type
+
+This proposal creates a new type of operator called `helm`. The new type is used to tell the tooling to act on that type of operator.
+
+### Package Structure
+
+Packages will be added to the operator-sdk. These packages are designed to be usable by the end user if they choose to and should have a well documented public API. The proposed packages are:
+
+* /operator-sdk/pkg/helm/client
+  * Will contain a helper function to create a Helm client from `controller-runtime` manager.
+
+* /operator-sdk/pkg/helm/controller
+  * Will contain the Helm controller.
+  * Will contain an exposed reconciler. The default `Add` method will use this reconciler.
+
+* /operator-sdk/pkg/helm/engine
+  * Will contain a Helm Engine implementation that adds owner references to generated Kubernetes resource assets, which is necessary for garbage collection of Helm chart resources.
+
+* /operator-sdk/pkg/helm/internal
+  * Will contain types and utilities used by other Helm packages in the SDK.
+
+* /operator-sdk/pkg/helm/release
+  * Will contain the Manager types and interfaces. A Manager is responsible for:
+    * Implementing Helm's Tiller functions that are necessary to install, update, and uninstall releases.
+    * Reconciling an existing release's resources.
+  * A default Manager implementation is provided in this package but is not exported.
+  * Package functions:
+    * NewManager - method that returns a new Manager for a provided helm chart.
+    * NewManagersFromEnv - method that returns a map of GVK to Manager types based on environment variables.
+    * NewManagersFromFile - method that returns a map of GVK to Manager types based on a provided config file.
+
+### Commands
+
+We are adding and updating existing commands to accommodate the helm operator.  Changes to the `cmd` package as well as changes to the generator are needed.
+
+#### New
+
+New functionality will be updates to allow Helm operator developers to create a new boilerplate operator structure with everything necessary to get started developing and deploying a Helm operator with the SDK.
+
+```
+operator-sdk new <project-name> --type=helm --kind=<kind> --api-version=<group/version>
+```
+
+This will be new scaffolding for the above command under the hood. We will:
+* Create a `./<project-name>` directory.
+* Create a `./<project-name>/charts` directory.
+* Generate a simple default chart at `./<project-name>/charts/<kind>`.
+* Create a new watches file at `./<project-name>/watches.yaml`. The chart and GVK will be defaulted based on input to the `new` command.
+* Create a `./<project-name>/deploy` with the Kubernetes resource files necessary to run the operator.
+* Create a `./build/Dockerfile` that uses the watches file and the helm chart. It will use the Helm operator as its base image.
+
+The resulting structure will be:
+
+```
+<project-name>
+|   watches.yaml
+|
+|-- build
+|   |   Dockerfile
+|
+|-- charts
```

Comment:
If I understood correctly `charts` here refers to collection of helm projects `<kind>` in this case. But in helm `charts` is reserved and expected to be a directory containing any charts upon which this chart depends. So I think this should be called something else to avoid confusion. WDYT?

Reply:
I found a few helm projects that refer to a collection of chart projects as just `charts`:
* https://github.com/helm/prow/blob/master/design.md#questions-and-answers
* https://github.com/helm/chartmuseum#api
* https://github.com/helm/charts

On the other hand, maybe this specific context of developing a Helm operator would make a user think the top-level project directory is the place to put their chart files? I'm not sure.

Would the fact that we scaffold out a boilerplate chart within the `charts` directory be enough of an indication to help users avoid that confusion?

Reply:
[Here](https://docs.helm.sh/developing_charts/#the-chart-file-structure) is where they mention the `charts` being reserved which made me think it might be confused for users.

> Would the fact that we scaffold out a boilerplate chart within the charts directory be enough of an indication to help users avoid that confusion?

 Yes that should be good enough, I guess if users face problems we can always adjust that in the future?

Reply:
Yeah I saw that. It's too bad the internal `charts` directory isn't named `dependencies` or similar. I'm not opposed to changing it now though. Maybe `helm-charts`? Any other suggestions?

Reply:
Changed to `helm-charts`